src/GitVersion.Core/Core/TaggedSemanticVersionService.cs

```diff
@@ -0,0 +1,346 @@
+using System.Collections.Concurrent;
+using GitVersion.Configuration;
+using GitVersion.Extensions;
+using GitVersion.Logging;
+
+namespace GitVersion.Core;
+
+internal interface IBranchRepository
+{
+    IEnumerable<IBranch> GetMainlineBranches(params IBranch[] excludeBranches);
+
+    IEnumerable<IBranch> GetReleaseBranches(params IBranch[] excludeBranches);
+}
+
+internal sealed class BranchRepository : IBranchRepository
+{
+    private GitVersionContext VersionContext => this.versionContextLazy.Value;
+    private readonly Lazy<GitVersionContext> versionContextLazy;
+
+    private readonly IGitRepository gitRepository;
+
+    public BranchRepository(Lazy<GitVersionContext> versionContext, IGitRepository gitRepository)
+    {
+        this.versionContextLazy = versionContext.NotNull();
+        this.gitRepository = gitRepository.NotNull();
+    }
+
+    public IEnumerable<IBranch> GetMainlineBranches(params IBranch[] excludeBranches)
+        => GetBranchesWhere(new HashSet<IBranch>(excludeBranches), configuration => configuration.IsMainline == true);
+
+    public IEnumerable<IBranch> GetReleaseBranches(params IBranch[] excludeBranches)
+        => GetBranchesWhere(new HashSet<IBranch>(excludeBranches), configuration => configuration.IsReleaseBranch == true);
+
+    private IEnumerable<IBranch> GetBranchesWhere(HashSet<IBranch> excludeBranches, Func<IBranchConfiguration, bool> predicate)
```

Comment:
`GetBranchesWhere` - I wonder if we can come up with a better name here

Reply:
Done